### PR TITLE
Various Set fixes and improvements

### PR DESCRIPTION
**Pull request by nirinchev**
## Description

* Adds `ISet<T>` support. Fixes https://github.com/realm/realm-dotnet/issues/2099.
* Adds tests for `ISet<RealmObject>` and `ISet<byte[]>`. Fixes https://github.com/realm/realm-dotnet/issues/2211.
* Wires up the native Set methods (currently only with a Set argument). Fixes https://github.com/realm/realm-dotnet/issues/2091
* Simplifies collection weaving by having RealmObjectBase.GetCollectionValue initialize the unmanaged collection.
* Implements `RealmObjectBase.GetHashCode`. Fixes https://github.com/realm/realm-dotnet/issues/1650.

##  TODO

* [ ] Changelog entry
* [x] Tests (if applicable)

### CODE REVIEW

**Review by papafe — APPROVED** [review]
Looks good to me

File: Realm/Realm.Fody/Common/RealmWeaver.cs [hunk 1]
```diff
@@ -405,32 +406,21 @@ private WeavePropertyResult WeaveProperty(PropertyDefinition prop, TypeDefinitio
...
-                        }
-
                         ReplaceCollectionGetter(prop, backingField, columnName,
-                                            new GenericInstanceMethod(_references.RealmObject_GetListValue) { GenericArguments = { elementType } },
-                                            concreteListConstructor);
+                                            new GenericInstanceMethod(_references.RealmObject_GetListValue) { GenericArguments = { elementType } });
                         break;
                     case RealmCollectionType.ISet:
-                        _logger.Warning($"{type.Name}.{prop.Name} is of type ISet which is not officially supported yet. Some functionality may not exist yet or there may be bugs/known issues that can result in undefined behavior, including data loss. Official support for the datatype will come in a future Realm release.");
+                        Debugger.Break();
```
Comment (LaPeste) [inline Realm/Realm.Fody/Common/RealmWeaver.cs:413]: Reminder to remove this when the PR is about to be merged.

File: Realm/Realm/DatabaseTypes/RealmObjectBase.cs [hunk 2]
```diff
@@ -453,4 +474,4 @@ public Metadata(TableKey tableKey, IRealmObjectHelper helper, IDictionary<string
             }
         }
     }
-}
\ No newline at end of file
+}
```
Comment (LaPeste) [inline Realm/Realm/DatabaseTypes/RealmObjectBase.cs:477]: I don't understand what the difference is here. The new line character or what?

File: Realm/Realm.Fody/Common/RealmWeaver.cs [hunk 3]
```diff
@@ -18,6 +18,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
```
Comment (LaPeste) [inline Realm/Realm.Fody/Common/RealmWeaver.cs:21]: And I guess the reminder applies here too.

File: Realm/Realm/DatabaseTypes/RealmCollectionBase.cs [hunk 4]
```diff
@@ -203,7 +203,7 @@ protected void AddToRealmIfNecessary(in RealmValue value)
         protected static EmbeddedObject EnsureUnmanagedEmbedded(in RealmValue value)
         {
             var result = value.AsRealmObject<EmbeddedObject>();
-            if (result.IsManaged)
+            if (result?.IsManaged == true)
```
Comment (LaPeste) [inline Realm/Realm/DatabaseTypes/RealmCollectionBase.cs:206]: ` == true` should not be necessary, why did you need it?

Comment (papafe) [reply Realm/Realm/DatabaseTypes/RealmCollectionBase.cs:206]: That's because `result?.IsManaged` is a nullable bool.

Comment (LaPeste) [reply Realm/Realm/DatabaseTypes/RealmCollectionBase.cs:206]: I educated myself. Makes sense.

**Review by LaPeste — COMMENTED** [review]
On `Tests/Realm.Tests/Database/RealmSetTests.cs` I went a little faster then the rest, but overall the changes look more then reasonable to me. However, I have a few doubts about the weaving part, so it could be beneficial if at the next PR reviews (Friday) you could spend a few words on the subject.

File: Realm/Realm.Fody/Common/RealmWeaver.cs (hunk 1 above)
Comment (nirinchev) [reply Realm/Realm.Fody/Common/RealmWeaver.cs:413]: 🤦‍♂️ Thanks for spotting it

File: Realm/Realm/DatabaseTypes/RealmObjectBase.cs (hunk 2 above)
Comment (nirinchev) [reply Realm/Realm/DatabaseTypes/RealmObjectBase.cs:477]: Yeah, the new editor config adds a new line at the end of the file. I made some changes to the file, then removed them, but am too lazy to revert the entire file (and the newline will inevitably be added eventually anyway).

**Review by LaPeste — APPROVED** [review]
